### ISSUE TITLE
[Proj 3] Runqueue Lock을 잡은 상태에서 `printk`할 때 데드락이 발생하는 경우

### ISSUE THREAD

**Issue by shp7724** [swsnu/osspr2022#109]
안녕하세요.

저희 팀에서 여러 load balancing interval을 테스트해보다가, runqueue lock을 잡은 상태에서 `printk`를 하는 경우 매우 드물게 데드락이 발생할 수 있다는 사실을 알게 되었는데요.
(`printk`->`vprintk_func`->`vprintk_default`->`vprintk_emit`->`console_unlock`->`__up_console_sem`->`up`->`__up`->`wake_up_process`->`try_to_wake_up`->`ttwu_queue`->`rq_lock`)

스펙에는 로드 밸런싱 로그를 `printk`로 출력하라는 내용이 있어서, 위와 같이 데드락이 발생하는 경우를 무시해도 되는 것인지, 아니면 printk variants나 기타 다른 방법을 사용해서 이러한 경우를 방지하도록 해야 하는 것인지(해도 되는 것인지) 궁금해서 질문드립니다.

감사합니다.

**Comment by WallabyG** [comment]
미리 필요한 값들을 읽어둔 뒤 rq lock을 풀고 printk를 사용하는 방법도 있을 것 같습니다.

**Comment by shp7724** [comment]
@WallabyG 네 그 방법도 고려하고 있긴 한데, 디버깅용으로 여기저기에 printk를 써놔서 아예 printk를 사용하지 않고 그 variant를 사용하는게 더 편할 것 같아서요. 스펙에 제시된 `printk` 대신 다른 함수를 사용해도 채점에 지장이 없을지가 궁금했습니다. 조언 감사드립니다!

**Comment by tirr-c** [comment]
스케줄러 코드에서는 `printk_deferred`를 쓰는 게 맞는 것 같습니다. 아마 채점에는 dmesg와 같은 커널 로그를 읽는 도구를 사용할 것 같아서, 이러한 printk variant를 사용하더라도 커널 로그에 남기만 하면 괜찮을 것 같습니다(제 생각입니다).

**Comment by widiba03304** [comment]
네 printk의 variants를 사용해도 괜찮습니다!
